src/lib.rs

```diff
@@ -0,0 +1,489 @@
+/// XXX This need to be user-specified, perhaps using a trait?
+type JitLocation = u32;
+
+/// The type used to count how many times a JitLocation has been interpreted.
+type HotCount = u32;
+
+/// The number of times a location is interpreted before we trace it.
+/// XXX make this configurable.
+const HOT_THRESHOLD: HotCount = 10;
+
+/// The shared trace store.
+///
+/// Maps a location to a trace (if one exists). Locations not in the map have not yet been traced.
+lazy_static! {
+    static ref TRACES: RwLock<HashMap<JitLocation, JitTrace>> = RwLock::new(HashMap::new());
+}
+
+/// Execute the trace for the specified location (if there is one).
+///
+/// Returns true if a trace was present (and thus was executed), or false otherwise.
+fn try_exec_trace(loc: JitLocation) -> bool {
+    let trace_map = TRACES.read().unwrap();
+    let trace = trace_map.get(&loc);
+    if trace == Some(&JitTrace::FinalisedTrace) {
+        JitThreadContext::phase_transition(JitPhase::RunningTrace(loc));
+        // XXX Really execute the trace.
+        JitThreadContext::phase_transition(JitPhase::Interpreting);
+        return true;
+    }
+    false
+}
+
+/// Is a thread currently tracing/compiling the specified location?
+fn trace_pending(loc: JitLocation) -> bool {
+    let trace_map = TRACES.read().unwrap();
+    match trace_map.get(&loc) {
+        Some(&JitTrace::TracePending) => return true,
+        _ => return false,
+    };
+}
+
+/// Store a trace (or "trace pending" marker in the trace cache for the specified location.
+fn set_trace(loc: JitLocation, trace: JitTrace) {
+    // Sanity checks.
+    {
+        // Extra scope for dropping the lock prior to the write() lock later (avoids deadlock).
+        let trace_map = TRACES.read().unwrap();
+        let old_val = trace_map.get(&loc);
+        if trace == JitTrace::TracePending {
+            // If we are collecting a trace, then the location should not already be in the trace cache.
+            assert!(old_val == None);
+        } else if trace == JitTrace::FinalisedTrace {
+            // If we are storing a finished trace, then it should have been marked as "pending" first.
+            assert!(old_val == Some(&JitTrace::TracePending));
+        }
+    }
+
+    TRACES.write().unwrap().insert(loc, trace);
+}
+
+/// Shared hot counts.
+///
+/// Maps a location to its hot count. Increments of the counter are atomic due to the RwLock.
+lazy_static! {
+    static ref HOT_COUNTS: RwLock<HashMap<JitLocation, HotCount>> = RwLock::new(HashMap::new());
+}
+
+/// Get the hot count for the given location.
+fn hot_count(loc: JitLocation) -> HotCount {
+    let ret = match HOT_COUNTS.read().unwrap().get(&loc) {
+        Some(v) => *v,
+        None => 0,
+    };
+    ret
+}
+
+/// Atomically increment and return the hot count for the given location.
+fn incr_hot_count(loc: JitLocation) -> HotCount {
+    // XXX don't store locations which cannot be loop heads.
```

Comment:
What is this XXX telling us? I think this is something that user interpreters should aim to do; it's not something the meta-tracer can do much about.

Reply:
This is the optimisation you suggested the other day. If it's something the interpreter must do, then yeah, let's just kill this comment.

Reply:
What I meant was it's something we should strongly encourage interpreter authors to do correctly.

Reply:
agreed